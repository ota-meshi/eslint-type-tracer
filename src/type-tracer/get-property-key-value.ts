import { getStaticValue } from "@eslint-community/eslint-utils";
import type { TSESTree } from "@typescript-eslint/types";
import type { Scope } from "eslint";

/**
 * Get the property name/symbol from a MemberExpression node or a Property node.
 *
 * The difference from `@eslint-community/eslint-utils.getPropertyName()` is
 * that if key is a Symbol, this function returns a Symbol.
 * @param {Property|MemberExpression} node The node to get.
 * @param {Scope} [initialScope] The scope to start finding variable. Optional. If the node is a computed property node and this scope was given, this checks the computed property name by the `getStringIfConstant` function with the scope, and returns the value of it.
 * @returns {string|symbol|null}
 */
export function getPropertyKeyValue(
  node:
    | TSESTree.Property
    | TSESTree.MemberExpression
    | TSESTree.MethodDefinition
    | TSESTree.PropertyDefinition,
  initialScope?: Scope.Scope,
): string | symbol | null {
  switch (node.type) {
    case "MemberExpression":
      if (node.computed) {
        return getStaticKeyValue(node.property, initialScope);
      }
      if (node.property.type === "PrivateIdentifier") {
        return null;
      }
      return node.property.name;

    case "Property":
    case "MethodDefinition":
    case "PropertyDefinition":
      if (node.computed) {
        return getStaticKeyValue(node.key, initialScope);
      }
      if (node.key.type === "Literal") {
        return String(node.key.value);
      }
      if (node.key.type === "PrivateIdentifier") {
        return null;
      }
      return node.key.name;

    // no default
  }

  return null;
}

/**
 * @param {Expression|PrivateIdentifier} node
 * @param {Scope} initialScope
 */
function getStaticKeyValue(
  node: TSESTree.Expression | TSESTree.PrivateIdentifier,
  initialScope?: Scope.Scope,
) {
  const value = getStaticValue(node, initialScope);
  return (
    value &&
    (typeof value.value === "symbol" ? value.value : String(value.value))
  );
}
