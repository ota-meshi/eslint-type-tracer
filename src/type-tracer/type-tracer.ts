import type { TypeChecker } from "./utils";
import type { SourceCode } from "eslint";
import { buildTypeCheckerForTS } from "./type-checker-for-ts";
import { buildTypeTracerForES } from "./type-tracer-for-es";

export type TypeCheckerOptions = {
  aggressiveResult?: false | "aggressive";
};

/**
 * Build object type checker.
 * @param context The rule context.
 * @param aggressiveResult The value to return if the type cannot be determined.
 * @returns Returns an object type checker.
 */
export function buildTypeChecker(
  sourceCode: SourceCode,
  options?: TypeCheckerOptions,
): TypeChecker {
  const aggressiveResult = options?.aggressiveResult ?? false;
  return (
    buildTypeCheckerForTS(sourceCode, aggressiveResult) ||
    buildTypeCheckerForES(sourceCode, aggressiveResult)
  );
}

/**
 * Build object type checker for TypeScript.
 */
function buildTypeCheckerForES(
  sourceCode: SourceCode,
  aggressiveResult: false | "aggressive",
): TypeChecker {
  const tracer = buildTypeTracerForES(sourceCode);
  return (node, className): boolean | "aggressive" => {
    const typeName = tracer(node);
    if (typeName == null) {
      return aggressiveResult;
    }
    return typeName === className;
  };
}
