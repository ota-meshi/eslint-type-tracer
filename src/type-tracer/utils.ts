import type { TSESTree } from "@typescript-eslint/types";
import type { TypeName } from "./types";

export type TypeTracer = (node: TSESTree.Expression) => TypeName[];

export type TypeChecker = (
  node: TSESTree.Expression,
  className: TypeName,
  memberAccessNode?: TSESTree.MemberExpression | TSESTree.Property,
) => boolean | "aggressive";

/**
 * Get the type of the given expression node.
 * @param node The expression node.
 * @returns `true` if should disallow it.
 */
export function getSimpleExpressionType(
  node: TSESTree.Expression | TSESTree.Super,
): TypeName | null {
  // If it's obvious, shortcut.
  if (node.type === "ArrayExpression") {
    return "Array";
  }
  if (node.type === "Literal") {
    if ("regex" in node && node.regex) {
      return "RegExp";
    }
    if ("bigint" in node && node.bigint) {
      return "BigInt";
    }
    if (node.value === null) {
      return "null";
    }
    if (typeof node.value === "string") {
      return "String";
    }
    if (typeof node.value === "number") {
      return "Number";
    }
    if (typeof node.value === "boolean") {
      return "Boolean";
    }
    if (typeof node.value === "bigint") {
      return "BigInt";
    }
    if (node.value instanceof RegExp) {
      return "RegExp";
    }
    return null;
  }
  if (node.type === "TemplateLiteral") {
    return "String";
  }
  if (node.type === "UpdateExpression") {
    return "Number";
  }

  return null;
}
