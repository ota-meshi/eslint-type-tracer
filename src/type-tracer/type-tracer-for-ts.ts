import type { TypeChecker, TypeTracer } from "./utils.ts";
import { getSimpleExpressionType } from "./utils.ts";
import * as module from "module";
import type * as typescript from "typescript";
import type { TypeName } from "./types.ts";
import type { TSESTree } from "@typescript-eslint/types";
import type { SourceCode } from "eslint";

let ts: typeof typescript;
try {
  const require = module.createRequire(import.meta.url);
  ts = require("typescript");
} catch {
  // ignore
}

/**
 * Build type tracer.
 * @returns Returns a type tracer.
 */
export function buildTypeTracerForTS(
  sourceCode: SourceCode,
): TypeTracer | null {
  const trace = buildTypeTracer<boolean>(sourceCode);
  if (!trace) return null;

  return function (node) {
    const type = getSimpleExpressionType(node);
    if (type) {
      return [type];
    }
    const result: TypeName[] = [];
    trace(node, {
      visitTypeName(typeName) {
        result.push(typeName as TypeName);
        return true;
      },
      visitUnknown() {
        return false;
      },
      visitSymbol(symbol, checker) {
        const name = checker.getFullyQualifiedName(symbol);
        if (!name) return false;
        if (name.startsWith("Readonly")) {
          result.push(name.slice(8) as TypeName);
        } else if (name === "CallableFunction") result.push("Function");
        else if (name === "IteratorObject") result.push("Iterator");
        else result.push(name as TypeName);
        return false;
      },
    });
    return result;
  };
}
/**
 * Build object type checker for TypeScript.
 * @param context The rule context.
 * @param aggressiveResult The value to return if the type cannot be determined.
 * @returns Returns an object type checker. Returns null if TypeScript is not available.
 */
export function buildTypeCheckerForTS(
  sourceCode: SourceCode,
  aggressiveResult: false | "aggressive" = false,
): TypeChecker | null {
  const trace = buildTypeTracer<boolean | "aggressive">(sourceCode);
  if (!trace) return null;

  return function (node, className, memberAccessNode) {
    const type = getSimpleExpressionType(node);
    if (type) {
      return type === className;
    }
    return trace(
      node,
      {
        visitTypeName(typeName) {
          return typeName === className;
        },
        visitUnknown() {
          return aggressiveResult;
        },
        visitSymbol(symbol, checker) {
          if (!className.includes(".")) {
            const escapedName = symbol.escapedName;
            return (
              escapedName === className ||
              // ReadonlyArray, ReadonlyMap, ReadonlySet
              escapedName === `Readonly${className}` ||
              // CallableFunction
              (className === "Function" &&
                escapedName === "CallableFunction") ||
              // IteratorObject
              (className === "Iterator" && escapedName === "IteratorObject")
            );
          }
          return checker.getFullyQualifiedName(symbol) === className;
        },
      },
      memberAccessNode,
    );
  };
}

interface Context<R> {
  visitSymbol(
    symbol: typescript.Symbol,
    checker: typescript.TypeChecker,
  ): R | false;
  visitUnknown(): R;
  visitTypeName(typeName: string): R;
}

type Trace<R> = (
  node: TSESTree.Expression,
  ctx: Context<R>,
  memberAccessNode?: TSESTree.MemberExpression | TSESTree.Property,
) => R | false;

/**
 *
 */
function buildTypeTracer<R>(sourceCode: SourceCode): Trace<R> | null {
  const tsNodeMap: ReadonlyMap<unknown, typescript.Node> =
    sourceCode.parserServices.esTreeNodeToTSNodeMap;
  const checker: typescript.TypeChecker =
    sourceCode.parserServices.program &&
    sourceCode.parserServices.program.getTypeChecker();

  const isTS = Boolean(ts && tsNodeMap && checker);
  if (!isTS) {
    return null;
  }
  const hasFullType =
    sourceCode.parserServices.hasFullTypeInformation !== false;

  return (node, ctx, memberAccessNode) => {
    return (
      (memberAccessNode && checkByPropertyDeclaration(memberAccessNode, ctx)) ||
      checkByObjectExpressionType(node, ctx)
    );
  };

  /**
   * Check if the type of the given node by the declaration of `node.property`.
   * @param memberAccessNode The MemberExpression or Property node.
   * @param ctx The trace context.
   * @returns a truthy value, if the process is complete.
   */
  function checkByPropertyDeclaration(
    memberAccessNode: TSESTree.MemberExpression | TSESTree.Property,
    ctx: Context<R>,
  ): R | false {
    const tsNode = tsNodeMap.get(
      memberAccessNode.type === "MemberExpression"
        ? memberAccessNode.property
        : memberAccessNode,
    );
    const symbol = tsNode && checker.getSymbolAtLocation(tsNode);
    const declarations = symbol && symbol.declarations;

    if (declarations) {
      for (const declaration of declarations) {
        if (declaration.parent.kind === ts.SyntaxKind.SourceFile) {
          continue;
        }
        const type = checker.getTypeAtLocation(declaration.parent);
        const r = type && visitType(type, ctx);
        if (r) {
          return r;
        }
      }
    }

    return false;
  }

  /**
   * Check if the type of the given node by the type of `node.object`.
   * @param node The Expression node.
   * @param ctx The trace context.
   * @returns a truthy value, if the process is complete.
   */
  function checkByObjectExpressionType(
    node: TSESTree.Expression,
    ctx: Context<R>,
  ): R | false {
    const tsNode = tsNodeMap.get(node)!;
    const type = checker.getTypeAtLocation(tsNode);
    return visitType(type, ctx);
  }

  /**
   * Visit the type.
   * @param type The type to check.
   * @param ctx The trace context.
   * @returns a truthy value, if the process is complete.
   */
  function visitType(type: typescript.Type, ctx: Context<R>): R | false {
    // console.log(
    //     "typeEquals(%o, %o)",
    //     {
    //         name: isClassOrInterface(type)
    //             ? type.symbol.escapedName
    //             : checker.typeToString(type),
    //         flags: Object.entries(ts.TypeFlags)
    //             .filter(
    //                 ([_id, flag]) =>
    //                     typeof flag === "number" &&
    //                     (type.flags & flag) === flag,
    //             )
    //             .map(([id]) => id)
    //             .join("|"),
    //         objectFlags:
    //             type.objectFlags == null
    //                 ? undefined
    //                 : Object.entries(ts.ObjectFlags)
    //                       .filter(
    //                           ([_id, flag]) =>
    //                               typeof flag === "number" &&
    //                               (type.objectFlags & flag) === flag,
    //                       )
    //                       .map(([id]) => id)
    //                       .join("|"),
    //         "symbol.flags": !type.symbol
    //             ? undefined
    //             : Object.entries(ts.SymbolFlags)
    //                   .filter(
    //                       ([_id, flag]) =>
    //                           typeof flag === "number" &&
    //                           (type.symbol.flags & flag) === flag,
    //                   )
    //                   .map(([id]) => id)
    //                   .join("|"),
    //     },
    //     className,
    // )
    if (isFunction(type)) {
      return ctx.visitTypeName("Function");
    }
    if (isAny(type) || isUnknown(type)) {
      return ctx.visitUnknown();
    }
    if (isAnonymousObject(type)) {
      // In non full-type mode, array types (e.g. `any[]`) become anonymous object type.
      return hasFullType ? false : ctx.visitUnknown();
    }

    if (isStringLike(type)) {
      return ctx.visitTypeName("String");
    }
    if (isNumberLike(type)) {
      return ctx.visitTypeName("Number");
    }
    if (isBooleanLike(type)) {
      return ctx.visitTypeName("Boolean");
    }
    if (isBigIntLike(type)) {
      return ctx.visitTypeName("BigInt");
    }
    if (isSymbolLike(type)) {
      return ctx.visitTypeName("Symbol");
    }
    if (isArrayLikeObject(type)) {
      return ctx.visitTypeName("Array");
    }

    if (isReferenceObject(type) && type.target !== type) {
      return visitType(type.target, ctx);
    }
    if (isTypeParameter(type)) {
      const constraintType = getConstraintType(type);
      if (constraintType) {
        return visitType(constraintType, ctx);
      }
      return hasFullType ? false : ctx.visitUnknown();
    }
    if (isUnionOrIntersection(type)) {
      for (const t of type.types) {
        const r = visitType(t, ctx);
        if (r) {
          return r;
        }
      }
      return false;
    }

    if (isClassOrInterface(type)) {
      return typeSymbolEscapedNameEquals(type, ctx);
    }
    return ctx.visitTypeName(checker.typeToString(type));
  }

  /**
   * Check if the symbol.escapedName of the given type is expected or not.
   * @param type The type to check.
   * @param ctx The trace context.
   * @returns a truthy value, if the process is complete.
   */
  function typeSymbolEscapedNameEquals(
    type: typescript.InterfaceType,
    ctx: Context<R>,
  ): R | false {
    const symbol = type.symbol;
    return ctx.visitSymbol(symbol, checker);
  }

  /**
   * Get the constraint type of a given type parameter type if exists.
   *
   * `type.getConstraint()` method doesn't return the constraint type of the
   * type parameter for some reason. So this gets the constraint type via AST.
   *
   * @param type The type parameter type to get.
   * @returns The constraint type.
   */
  function getConstraintType(type: typescript.TypeParameter) {
    const declaration = type.symbol?.declarations?.[0];
    if (
      declaration &&
      ts.isTypeParameterDeclaration(declaration) &&
      declaration.constraint != null
    ) {
      return checker.getTypeFromTypeNode(declaration.constraint);
    }
    return undefined;
  }
}

/**
 * Check if a given type is an anonymous object type or not.
 * @param type The type to check.
 * @returns `true` if the type is an anonymous object type.
 */
function isAnonymousObject(type: typescript.Type) {
  return isObject(type) && (type.objectFlags & ts.ObjectFlags.Anonymous) !== 0;
}

/**
 * Check if a given type is `any` or not.
 * @param type The type to check.
 * @returns `true` if the type is `any`.
 */
function isAny(type: typescript.Type) {
  return (type.flags & ts.TypeFlags.Any) !== 0;
}

/**
 * Check if a given type is an array-like type or not.
 * @param type The type to check.
 * @returns `true` if the type is an array-like type.
 */
function isArrayLikeObject(type: typescript.Type) {
  return (
    isObject(type) &&
    (type.objectFlags &
      (ts.ObjectFlags.ArrayLiteral |
        ts.ObjectFlags.EvolvingArray |
        ts.ObjectFlags.Tuple)) !==
      0
  );
}

/**
 * Check if a given type is an interface type or not.
 * @param type The type to check.
 * @returns `true` if the type is an interface type.
 */
function isClassOrInterface(
  type: typescript.Type,
): type is typescript.InterfaceType {
  return (
    isObject(type) && (type.objectFlags & ts.ObjectFlags.ClassOrInterface) !== 0
  );
}

/**
 * Check if a given type is an object type or not.
 * @param type The type to check.
 * @returns `true` if the type is an object type.
 */
function isObject(type: typescript.Type): type is typescript.ObjectType {
  return (type.flags & ts.TypeFlags.Object) !== 0;
}

/**
 * Check if a given type is a reference type or not.
 * @param type The type to check.
 * @returns `true` if the type is a reference type.
 */
function isReferenceObject(
  type: typescript.Type,
): type is typescript.TypeReference {
  return isObject(type) && (type.objectFlags & ts.ObjectFlags.Reference) !== 0;
}

/**
 * Check if a given type is a string-like type or not.
 * @param type The type to check.
 * @returns `true` if the type is a string-like type.
 */
function isStringLike(type: typescript.Type) {
  return (type.flags & ts.TypeFlags.StringLike) !== 0;
}

/**
 * Check if a given type is a number-like type or not.
 * @param type The type to check.
 * @returns `true` if the type is a number-like type.
 */
function isNumberLike(type: typescript.Type) {
  return (type.flags & ts.TypeFlags.NumberLike) !== 0;
}

/**
 * Check if a given type is a boolean-like type or not.
 * @param type The type to check.
 * @returns `true` if the type is a boolean-like type.
 */
function isBooleanLike(type: typescript.Type) {
  return (type.flags & ts.TypeFlags.BooleanLike) !== 0;
}

/**
 * Check if a given type is a bigint-like type or not.
 * @param type The type to check.
 * @returns `true` if the type is a bigint-like type.
 */
function isBigIntLike(type: typescript.Type) {
  return (type.flags & ts.TypeFlags.BigIntLike) !== 0;
}

/**
 * Check if a given type is a symbol-like type or not.
 * @param type The type to check.
 * @returns `true` if the type is a symbol-like type.
 */
function isSymbolLike(type: typescript.Type) {
  return (type.flags & ts.TypeFlags.ESSymbolLike) !== 0;
}

/**
 * Check if a given type is a type parameter type or not.
 * @param type The type to check.
 * @returns `true` if the type is a type parameter type.
 */
function isTypeParameter(type: typescript.Type) {
  return (type.flags & ts.TypeFlags.TypeParameter) !== 0;
}

/**
 * Check if a given type is a union-or-intersection type or not.
 * @param type The type to check.
 * @returns `true` if the type is a union-or-intersection type.
 */
function isUnionOrIntersection(
  type: typescript.Type,
): type is typescript.UnionOrIntersectionType {
  return (type.flags & ts.TypeFlags.UnionOrIntersection) !== 0;
}

/**
 * Check if a given type is `unknown` or not.
 * @param type The type to check.
 * @returns `true` if the type is `unknown`.
 */
function isUnknown(type: typescript.Type) {
  return (type.flags & ts.TypeFlags.Unknown) !== 0;
}

/**
 * Check if a given type is `function` or not.
 * @param type The type to check.
 * @returns `true` if the type is `function`.
 */
function isFunction(type: typescript.Type) {
  if (
    type.symbol &&
    (type.symbol.flags & (ts.SymbolFlags.Function | ts.SymbolFlags.Method)) !==
      0
  ) {
    return true;
  }

  const signatures = type.getCallSignatures();
  return signatures.length > 0;
}
