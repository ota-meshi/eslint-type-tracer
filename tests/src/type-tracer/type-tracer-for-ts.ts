import { deepStrictEqual } from "assert";
import { buildTypeTracerForTS } from "../../../src/type-tracer/type-tracer-for-ts";
import { expect } from "@ota-meshi/test-snapshot";
import type { TypeName } from "../../../src/type-tracer/types";
import type { Rule } from "eslint";
import { Linter } from "eslint";
import { fileURLToPath } from "url";
import path from "path";
import fs from "fs";
import * as tsParser from "@typescript-eslint/parser";

const FIXTURES_ROOT = path.resolve(
  fileURLToPath(import.meta.url),
  "../../../fixtures/ts",
);

function* extractFixtures(dir = FIXTURES_ROOT): Generator<{
  name: string;
  code: string;
  filename: string;
  only?: boolean;
}> {
  for (const dirent of fs.readdirSync(dir, { withFileTypes: true })) {
    const filePath = path.join(dir, dirent.name);
    if (dirent.isFile()) {
      if (dirent.name.endsWith(".ts")) {
        yield {
          name: path.relative(FIXTURES_ROOT, filePath),
          code: fs.readFileSync(filePath, "utf8"),
          filename: filePath,
        };
      }
    } else if (dirent.isDirectory()) {
      yield* extractFixtures(filePath);
    }
  }
}

describe("type-tracer-for-ts", () => {
  describe("buildTypeTracerForTS", () => {
    for (const { name, code, filename, only } of extractFixtures()) {
      describe(name, () => {
        (only ? it.only : it)(code, () => {
          expect(
            getResultOfBuildTypeTracerForTS(code, filename),
          ).toMatchSnapshot();
        });
      });
    }
  });
});

function getResultOfBuildTypeTracerForTS(code: string, filename: string) {
  const linter = new Linter({ configType: "flat" });

  const result: {
    expr: string;
    types: (TypeName | null)[];
  }[] = [];
  const linterResult = linter.verify(
    code,
    {
      files: ["**/*.*"],
      plugins: {
        test: {
          rules: {
            "test-rule": {
              create(context: Rule.RuleContext) {
                const getType = buildTypeTracerForTS(context.sourceCode);
                return {
                  "CallExpression[callee.name = target]"(node) {
                    result.push({
                      expr: context.sourceCode.getText(node),
                      types: node.arguments.map(getType),
                    });
                  },
                };
              },
            },
          },
        },
      },
      languageOptions: {
        ecmaVersion: "latest",
        sourceType: "module",
        parser: tsParser,
        parserOptions: {
          tsconfigRootDir: FIXTURES_ROOT,
          project: "tsconfig.test.json",
        },
      },
      rules: { "test/test-rule": "warn" },
    },
    {
      filename,
    },
  );
  deepStrictEqual(linterResult, []);
  return result;
}
