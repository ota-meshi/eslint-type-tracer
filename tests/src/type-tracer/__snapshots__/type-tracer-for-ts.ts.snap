// test-snapshot v1

exports[`type-tracer-for-ts buildTypeTracerForTS array.ts import { target } from "./target";

const a = [1, 2, 3, 4, 5];

target([], a);

function fn<X extends any[]>(
  a: readonly number[],
  b: readonly number[],
  c: X,
): void {
  target(a, b, c);
}
 1`] = `
Array [
  Object {
    "expr": "target([], a)",
    "types": Array [
      "Array",
      "Array",
    ],
  },
  Object {
    "expr": "target(a, b, c)",
    "types": Array [
      "Array",
      "Array",
      "Array",
    ],
  },
]
`;

exports[`type-tracer-for-ts buildTypeTracerForTS function.ts import { target } from "./target";

target(fn, String);

function fn<X extends (...arg: any[]) => void>(
  a: { (a: any): void },
  b: () => unknown,
  c: X,
): void {
  target(a, b, c);
}
 1`] = `
Array [
  Object {
    "expr": "target(fn, String)",
    "types": Array [
      "Function",
      "Function",
    ],
  },
  Object {
    "expr": "target(a, b, c)",
    "types": Array [
      "Function",
      "Function",
      "Function",
    ],
  },
]
`;

exports[`type-tracer-for-ts buildTypeTracerForTS target.ts export function target(...args: any[]): void {
  throw new Error("For Test");
}
 1`] = `Array []`;
